packages/sitecore-jss/src/personalization/layout-personalization-service.ts

```diff
@@ -0,0 +1,220 @@
+      },
+      personalizedRenderings
+    );
+
+    try {
+      const components = await currentResult.personalizationOperation;
+      currentResult.components = components;
+      return { isPersonalizable: true };
+    } catch (error) {
+      currentResult.personalizationOperation = undefined;
+      throw error;
+    }
+  }
+
+  /**
+   * Gets the personalized component.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  getPersonalizedComponent(componentUid: string): ComponentRendering | null {
+    if (!this.personalizationResult.components) {
+      return null;
+    }
+
+    return this.personalizationResult.components[componentUid] ?? null;
+  }
+
+  /**
+   * Provides a value that indicates whether the loading is in-progress.
+   * @returns {boolean} The value that indicates whether the loading is in-progress.
+   */
+  isLoading(): boolean {
+    return (
+      !!this.personalizationResult.personalizationOperation &&
+      !this.personalizationResult.components
+    );
+  }
+
+  /**
+   * Ensures the personalized component is loaded.
+   * @param {string} componentUid The unique identifier of a component.
+   * @returns {ComponentRendering} The personalized component.
+   */
+  async ensurePersonalizedComponentLoaded(
+    componentUid: string
+  ): Promise<ComponentRendering | null> {
+    if (!this.personalizationResult.personalizationOperation) {
+      throw new Error(
+        `${this.fetchPersonalization.name} should be called before getting personalized component`
+      );
+    }
+
+    const personalizedComponents = await this.personalizationResult.personalizationOperation;
+    if (!personalizedComponents) {
+      return null;
+    }
+
+    return personalizedComponents[componentUid] ?? null;
+  }
+
+  /**
+   * Personalizes components.
+   * @param {PersonalizationContext} context The context.
+   * @param {PersonalizableComponentRendering[]} personalizableRenderings The personalizable components.
+   * @returns {Object.<string, ComponentRendering | null>} The personalized components.
+   */
+  async personalizeComponents(
+    context: PersonalizationContext,
+    personalizableRenderings: PersonalizableComponentRendering[]
+  ): Promise<{ [key: string]: ComponentRendering | null }> {
+    if (personalizableRenderings.length === 0) {
+      return {};
+    }
+
+    const personalizedRenderingIds = personalizableRenderings.map((r) => r.uid);
+    let personalizedFragments: { [key: string]: ComponentRendering | null | undefined } = {};
+
+    try {
+      const personalizationDecisionsResult = await this.personalizationDecisionsService.getPersonalizationDecisions(
+        {
+          routePath: context.routePath,
+          language: context.language,
+          renderingIds: personalizedRenderingIds,
+        }
+      );
+      personalizedFragments = await this.resolveFragments(personalizationDecisionsResult, context);
```

Comment:
Instead of 2 awaits, this could utilize Promise.then

Instead of wrapping in a try/catch, this could utilize Promise.catch

Reply:
We had Promise.then/catch before but Illia asked us to replace with async/await. :-)